### PR TITLE
[Core] GC dead nodes reported from autoscaler

### LINKED ISSUE #35874: [core][autoscaler] Autoscaler V2: GC dead nodes
### What happened + What you expected to happen

We should have a GC policy in case of a long running cluster with many dead nodes. 

### Versions / Dependencies

NA

### Reproduction script

NA

### Issue Severity

None

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

Dead nodes are actually already GC'ed when it exceeds 1000. https://github.com/ray-project/ray/blob/8a1fe6e69fdeb98bb220e4e19f60c2078b5e057a/src/ray/common/ray_config_def.h#L385

We have 2 options to fix the issue. 1. we reduce this. 2. we report only the subset of dead nodes. 

I chose the second option. It also reports the most recently dead nodes.

## Related issue number

Closes https://github.com/ray-project/ray/issues/35874

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/common/ray_config_def.h [hunk 1]
```diff
@@ -891,3 +891,7 @@ RAY_CONFIG(int, object_manager_rpc_threads_num, 0)
 
 // Write export API events to file if enabled
 RAY_CONFIG(bool, enable_export_api_write, false)
+
+// Only applicable when enable_autoscaler_v2 is enabled.
+// The maximum number of dead nodes to report from the autoscaler.
+RAY_CONFIG(uint32_t, autoscaler_maximum_gcs_dead_node_to_report, 100)
```
Comment (rkooo567) [inline src/ray/common/ray_config_def.h:897]: this is an arbitrary value. We can discuss what's the best one

**Comment by rkooo567** [comment]
yeah that's possible. or we can reduce the dead node cache size smaller. I am open to all solutions.

**Comment by rynewang** [comment]
why don't we just reduce the number, if the mechanism is already there?

**Comment by rkooo567** [comment]
2 reasons. But the reasons are not very strong;
- it is backward incompatible changes
- dahsboard will lose access to dead nodes earlier. 

it is the best way to keep the existing behavior while we reduce the payload size reported, but I am okay with wahtever tbh. I think we should just let the prod team decide

**Comment by rkooo567** [comment]
Talked with @shomilj. For now, they don't have clear evidence if dead nodes are the root cause if the issue. And likely 1000  nodes limitation (the default) is okay. They will do more scale testing and deatiled analysis in the future and we can revisit the approach. I will close both issues for now

**Pull request closed without merging**